### DISCUSSION TITLE
Does `internalChecksFilter` have a higher priority than the `stabilityDays & automerge` combination?

### DISCUSSION THREAD

**Discussion by bahrmichael** [renovatebot/renovate#17113] [Request Help]
There are three parameters `stabilityDays`, `internalChecksFilter`, and `automerge` which seem to influence when Renovate creates a PR, and if it will create it before it can be merged or not.

We have found two paragraphs in the documentation that address how quickly PRs are created, and if they stay open for a while:
- https://docs.renovatebot.com/configuration-options/#suppress-branchpr-creation-for-x-days
- https://docs.renovatebot.com/configuration-options/#await-x-days-before-automerging

We're currently unsure if the combination `stabilityDays=21 & internalChecksFilter="strict" & automerge=true` will create PRs and keep them open until 21 days have passed, or if it would wait until 21 days have passed before a PR is created.

**So the main question is**: Does `internalChecksFilter` have a higher priority than the `stabilityDays & automerge` combination?

In our ideal world we would like to receive a pull request 21 days after a release is published (to reduce the risk of supply chain attacks), and want to have PRs open for the least amount of time. Ideally a PR is opened, and 15 minutes later it's in the main branch.

I have seen the discussions https://github.com/renovatebot/renovate/discussions/10419 and https://github.com/renovatebot/renovate/discussions/16688, but they don't involve the `automerge` property. Hence I'm asking here :)

**Comment by rarkins** [comment]
I think you'll find that `internalChecksFilter=strict` means you get no branch or PR until the stabilityDays is satisfied

**Reply by bahrmichael** [reply]
We have one repository where we get immediate PRs, and one where we get them when the stabilityDays are reached. I've not found a difference between those. Do you have an idea of what the reason could be? Any other information that's helpful to understanding the issue?

My understanding is that the differences below should only affect lockFileMaintenance and GitHub workflows, but not regular upgrades.

**gets immediate PRs**

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>MyOrg/renovate-config",
    "helpers:pinGitHubActionDigests",
  ],
  "lockFileMaintenance": {
    "enabled": true,
    "schedule": ["at any time"],
    "automerge": true,
    "platformAutomerge": true,
  }
}
```

**gets PRs when stabilityDays reached**

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>MyOrg/renovate-config"
  ]
}
```


**central config**

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "stabilityDays": 21,
  "internalChecksFilter": "strict",
  "major": {
    // ...
  },
  "replacement": {
    // ...
  },
  "vulnerabilityAlerts": {
    // ...
  },
  "lockFileMaintenance": {
    "enabled": true,
    "schedule": [
      "at any time"
    ],
  },
  "packageRules": [
    // ...
    {
      "matchUpdateTypes": [
        "minor",
        "patch",
        "pin",
        "digest",
        "pinDigest",
        "lockFileMaintenance",
        "rollback",
        "bump"
      ],
      "matchCurrentVersion": "!/^0/",
      "automerge": true,
      "platformAutomerge": true
    }
  ]
}
```

**Reply by viceice** [reply]
it depends on the datasource and registry, not all of them are returning the required release time stamp, so `stabilityDays` won't work.

**Reply by bahrmichael** [reply]
The PRs we're looking at come from npmjs.com which should be good, right? Or can this data be missing on a package level?

**Reply by bahrmichael** [reply]
In our case we see PRs where Renovate is waiting for stabilityDays to complete. My understanding is that it has enough information to eventually succeed on that check. Is that right? Could Renovate say it's waiting for the stabilityDays, but never succeed because it doesn't have the info to do so?

**Reply by viceice** [reply]
do you use a npm proxy like proget? that's known to have issues with release timestamps

**Reply by rarkins** [reply]
lock file maintenance has no concept of stabilityDays

**Reply by siddharthsarda** [reply]
> lock file maintenance has no concept of stabilityDays

@rarkins why is that?

**Reply by rarkins** [reply]
It's practically impossible and be pointless even if it was possible. 

The package manager itself will be used to generate a full refresh of the lock file. Potentially hundreds of dependencies may be updated in this process and some of them are bound to be very new so the check may never pass even if we had the capability to reverse engineer all release dates
